In [1]:
!pip install pandas numpy matplotlib seaborn scikit-learn

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

df = pd.read_csv('../data/raw/FinPay_input.csv')

print("Datasetin ölçüsü (sətir, sütun):", df.shape)
display(df.head())
df.info()
display(df.describe(include='all'))


[notice] A new release of pip is available: 24.3.1 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


Datasetin ölçüsü (sətir, sütun): (18072, 24)


,customer_id,age,gender,country,monthly_income,account_balance,credit_score,months_with_company,transactions_last_month,avg_transaction_value,...,app_logins,days_since_last_login,support_tickets,failed_transactions,card_usage_frequency,international_transactions,cashback_received,marketing_email_clicks,next_month_spending,churn
0,FP100000,26.0,Female,France,1865.18,0.00,485.0,101.0,29.0,80.90,...,19.0,1.0,0.0,3.0,0.898,1.0,31.51,3.0,1354.87,0
1,FP100001,55.0,Male,Germany,5367.68,4957.19,718.0,31.0,24.0,92.42,...,17.0,7.0,0.0,2.0,0.893,2.0,5.20,0.0,2302.22,0
2,FP100002,34.0,Male,France,2546.21,85.85,661.0,41.0,27.0,198.24,...,24.0,12.0,1.0,0.0,0.957,2.0,67.30,6.0,2750.36,0
3,FP100003,37.0,Female,United Kingdom,6186.12,10317.05,605.0,25.0,24.0,41.93,...,18.0,2.0,0.0,2.0,0.783,0.0,3.73,3.0,1831.61,0
4,FP100004,33.0,Female,Ireland,4874.31,10450.13,683.0,10.0,18.0,184.89,...,12.0,1.0,0.0,0.0,0.546,5.0,36.90,4.0,2448.02,1


<class 'pandas.DataFrame'>
RangeIndex: 18072 entries, 0 to 18071
Data columns (total 24 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   customer_id                 18072 non-null  str    
 1   age                         18072 non-null  float64
 2   gender                      17944 non-null  str    
 3   country                     17964 non-null  str    
 4   monthly_income              17747 non-null  float64
 5   account_balance             17856 non-null  float64
 6   credit_score                17819 non-null  float64
 7   months_with_company         18072 non-null  float64
 8   transactions_last_month     18072 non-null  float64
 9   avg_transaction_value       17964 non-null  float64
 10  monthly_spending            18072 non-null  float64
 11  previous_month_spending     18072 non-null  float64
 12  savings_amount              18072 non-null  float64
 13  loan_amount                 18072 non-null

,customer_id,age,gender,country,monthly_income,account_balance,credit_score,months_with_company,transactions_last_month,avg_transaction_value,...,app_logins,days_since_last_login,support_tickets,failed_transactions,card_usage_frequency,international_transactions,cashback_received,marketing_email_clicks,next_month_spending,churn
count,18072,18072.000000,17944,17964,17747.000000,17856.000000,17819.000000,18072.000000,18072.000000,17964.000000,...,17910.000000,18072.000000,18072.000000,18072.000000,18072.000000,18072.000000,18072.000000,18072.000000,18072.000000,18072.000000
unique,18000,NaN,8,12,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
top,FP100545,NaN,Female,United Kingdom,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
freq,2,NaN,8573,5390,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
mean,NaN,39.191124,NaN,NaN,4143.688506,9479.861536,675.398732,41.222554,21.156153,67.054633,...,18.193970,7.261178,0.686864,0.463701,0.708575,2.193891,11.373048,3.411797,1668.602707,0.107016
std,NaN,11.748433,NaN,NaN,2740.567608,10194.714773,72.046479,26.520637,4.952295,39.828402,...,4.204939,10.962216,0.843224,0.702580,0.173303,1.525295,9.542616,1.852114,752.384879,0.309143
min,NaN,16.000000,NaN,NaN,-500.000000,0.000000,120.000000,1.000000,3.000000,6.110000,...,4.000000,-3.000000,0.000000,0.000000,0.091000,0.000000,0.210000,0.000000,50.000000,0.000000
25%,NaN,31.000000,NaN,NaN,2533.345000,4136.502500,627.000000,22.000000,18.000000,39.940000,...,15.000000,2.000000,0.000000,0.000000,0.587000,1.000000,5.010000,2.000000,1183.885000,0.000000
50%,NaN,39.000000,NaN,NaN,3578.590000,7605.125000,676.000000,36.000000,21.000000,57.470000,...,18.000000,5.000000,0.000000,0.000000,0.698000,2.000000,8.710000,3.000000,1576.060000,0.000000
75%,NaN,47.000000,NaN,NaN,5095.120000,12370.230000,723.000000,55.000000,24.000000,83.162500,...,21.000000,10.000000,1.000000,1.000000,0.820000,3.000000,14.690000,5.000000,2031.752500,0.000000


In [2]:
missing_df = pd.DataFrame({
    'Bos dəyərlərin sayı': df.isnull().sum(),
    'Faiz (%)': (df.isnull().sum() / len(df)) * 100
})
print("--{--- Boş dəyərlərin sayı ---}--")
display(missing_df[missing_df['Bos dəyərlərin sayı'] > 0])

tekrarlar = df.duplicated().sum()
print(f"\nTekrar setirlerin sayi:{tekrarlar}")

--{--- Boş dəyərlərin sayı ---}--


,Bos dəyərlərin sayı,Faiz (%)
gender,128,0.708278
country,108,0.597610
monthly_income,325,1.798362
account_balance,216,1.195219
credit_score,253,1.399956
avg_transaction_value,108,0.597610
app_logins,162,0.896414



Tekrar setirlerin sayi:72


In [3]:
cat_unique = ['gender','country']

for col in cat_unique:
    print(f"\nSütun: {col}")
    print(df[col].value_counts(dropna=False))
    print("-" * 40)


Sütun: gender
gender
Female               8573
Male                 8330
Prefer not to say     697
Non-binary            272
NaN                   128
male                   19
F                      19
female                 18
M                      16
Name: count, dtype: int64
----------------------------------------

Sütun: country
country
United Kingdom    5390
Germany           3578
France            2301
Italy             2015
Spain             1930
Netherlands       1606
Ireland           1049
NaN                108
U.K.                27
DE                  22
UK                  18
United kingdom      18
germany             10
Name: count, dtype: int64
----------------------------------------


In [4]:
# Klassifikasiya
print("=== TARGET 1: CHURN PAYLANMASI (%) ===")
print(df['churn'].value_counts(normalize=True) * 100)

print("\n" + "="*40 + "\n")

# regression
print("=== TARGET 2: NEXT_MONTH_SPENDING İSTATİSTİKLERİ ===")
print(df['next_month_spending'].describe())

=== TARGET 1: CHURN PAYLANMASI (%) ===
churn
0    89.298362
1    10.701638
Name: proportion, dtype: float64


=== TARGET 2: NEXT_MONTH_SPENDING İSTATİSTİKLERİ ===
count    18072.000000
mean      1668.602707
std        752.384879
min         50.000000
25%       1183.885000
50%       1576.060000
75%       2031.752500
max      14266.365138
Name: next_month_spending, dtype: float64


Faza 1 – Data Understanding Yekun Müşahidələri:
1. Dataset Həcmi: Dataset 18,072 sətir və 24 sütundan ibarətdir.
2. Çatışmayan Dəyərlər (Missing Values): "gender" (128), "country" (108), "monthly_income" (325) daxil olmaqla 7 sütunda az miqdarda eksik dəyər mövcuddur.
3. Kategorik Dəyərlərdə Uyğunsuzluq: "gender" sütununda eyni anlayışı bildirən müxtəlif yazılış formaları ("Female", "female", "F", "Male", "male", "M") var və standartlaşdırılmalıdır.
4. Dublikatlar və Anomaliyalar: Datasetdə 72 ədəd dublikat sətir tapılmışdır. Həmçinin "age" sütununda maks dəyərin 120 olması kimi anomaliyalar təmizlənməlidir.
5. Hədəf Dəyişənlərinin (Targets) Paylanması: "churn" sütununda böyük sinif balanssızlığı mövcuddur (%89.3-ə %10.7). "next_month_spending" göstəricisində isə maks dəyərin (14,266) yüksək olması outlierlerin cox olduğunu göstərir.